Docs: resource group modifications (#16321)
greenplum-db/gpdb: gpdb-doc/markdown/admin_guide/workload_mgmt_resgroups.html.md

title: Using Resource Groups 
---

> You use resource groups to set and enforce CPU, memory, concurrent transaction limits, and disk I/O in Greenplum Database. Once you define a resource group, you assign the group to one or more Greenplum Database roles in order to control the resources used by them. 

Reviewer: The first sentence I am thinking more like:

```
You use resource groups to manage and protect the resources allocations like CPU, memory, IO (footnote: cgroup V2 only), and transaction concurrency limit in Greenplum database.
```

|Limit Type|Description|Value Range|Default|
|----------|-----------|-----| ------| 

> |CONCURRENCY|The maximum number of concurrent transactions, including active and idle transactions, that are permitted in the resource group.| [0-[max_connections](../ref_guide/config_params/guc-list.html#max_connections)] | 20 |

Reviewer: ```suggestion
|CONCURRENCY|The maximum number of concurrent transactions, including active and idle transactions, that are permitted in the resource group.| [0 - [max_connections](../ref_guide/config_params/guc-list.html#max_connections)] | 20 |
```

> |CPU_MAX_PERCENT|The maximum percentage of CPU resources the group can use.| [1-100] | -1 (not set)|

Reviewer: ```suggestion
|CPU_MAX_PERCENT|The maximum percentage of CPU resources the group can use.| [1 - 100] | -1 (not set)|
```

> |CPU_WEIGHT|The scheduling priority of the resource group.| [1-500] | 100 |

Reviewer: ```suggestion
|CPU_WEIGHT|The scheduling priority of the resource group.| [1 - 500] | 100 |
```

|Limit Type|Description|Value Range|Default|
|----------|-----------|-----| ------| 
|CONCURRENCY|The maximum number of concurrent transactions, including active and idle transactions, that are permitted in the resource group.| 0-[max_connections](../ref_guide/config_params/guc-list.html#max_connections) | 20 |
|CPU_MAX_PERCENT|The maximum percentage of CPU resources the group can use.| [1-100] | -1 (not set)|
|CPU_WEIGHT|The scheduling priority of the resource group.| [1-500] | 100 |
|CPUSET|The specific CPU logical core (or logical thread in hyperthreading) reserved for this resource group.| It depends on system core configuration | -1 |

> |IO_LIMIT| The limit for the maximum read/write disk I/O throughput, and maximum read/write I/O operations per second. Set the value on a per-tablespace basis.| [2 - 4294967295 or `max`] | -1 |

Author: @wenwangBJ @SmartKeyerror

> |IO_LIMIT| The limit for the maximum read/write disk I/O throughput, and maximum read/write I/O operations per second. Set the value on a per-tablespace basis.| [2 - 4294967295 or `max`] | -1 |

Reviewer: I saw the example later on. So, if we can get confirmation from the server team, then this can be resolved.

> |IO_LIMIT| The limit for the maximum read/write disk I/O throughput, and maximum read/write I/O operations per second. Set the value on a per-tablespace basis.| [2 - 4294967295 or `max`] | -1 |

Reviewer 2: we do not design as that, so need to specify all the 4 options.

> |IO_LIMIT| The limit for the maximum read/write disk I/O throughput, and maximum read/write I/O operations per second. Set the value on a per-tablespace basis.| [2 - 4294967295 or `max`] | -1 |

Reviewer 3: @xinzweb Yes, we can.

These sentences have the same meaning:
-  alter resource group rg1 set io_limit '1663:rbps=max';
- alter resource group rg1 set io_limit '1663:rbps=max,wbps=max,riops=max,wiops=max';

Another example, these sentences have the same meaning:
- create resource group rg5 with (concurrency=10, cpu_max_percent=10, io_limit='*:rbps=1000');
- create resource group rg5 with (concurrency=10, cpu_max_percent=10, io_limit='*:rbps=1000,,wbps=max,riops=max,wiops=max');

> |IO_LIMIT| The limit for the maximum read/write disk I/O throughput, and maximum read/write I/O operations per second. Set the value on a per-tablespace basis.| [[2 - 4294967295 or `max`] | -1 |

Reviewer 3: If the parameter IO_LIMIT is not set, the default value for rbps, wpbs, riops, and wiopss is set to max, which means that there are no disk I/O limits. In this scenario, the gp_toolkit.gp_resgroup_config system view displays its value as -1.

The parameter `CPU_MAX_PERCENT` reserves the specified percentage of the segment CPU for resource management. The minimum `CPU_MAX_PERCENT` percentage you can specify for a resource group is 1, the maximum is 100. The sum of `CPU_MAX_PERCENT`s specified for all resource groups that you define in your Greenplum Database cluster can exceed 100. It specifies the total time ratio that all tasks in a resource group can run in one CPU cycle. Once the tasks in the resource group have used up all the time specified by the quota, they are throttled for the remainder of the time specified in that time period, and are not allowed to run until the next time period. 

When tasks in a resource group are idle and not using any CPU time, the leftover time is collected in a global pool of unused CPU cycles. Other resource groups can borrow CPU cycles from this pool. The actual amount of CPU time available to a resource group may vary, depending on the number of resource groups present on the system.

> The parameter `CPU_MAX_LIMIT` enforces a hard upper limit of CPU usage. For example, if it is set to 40%, it indicates that although the resource group can temporarily use some idle CPU resources from other groups, the maximum it can use is 40% of the CPU resources available to Greenplum. 

Reviewer 3: Change "CPU_MAX_LIMIT" to "CPU_MAX_PERCENT".

Reviewer 3: I didn't see this change in this PR, please update:
Change "CPU_MAX_LIMIT" to "CPU_MAX_PERCENT". We don't have "CPU_MAX_LIMIT"  in GP7 any more.

- The maximum value of `statement_mem` is capped at [max_statement_mem](../ref_guide/config_params/guc-list.html#max_statement_mem).
- Queries whose plan cost is less than the limit `MIN_COST` use a memory quota of `statement_mem`. 

### <a id="diskio"></a>Disk I/O Limits

Greenplum Database leverages Linux control groups to implement disk I/O limits. The parameter `IO_LIMIT` limits the maximum read/write disk I/O throughput, and the maximum read/write I/O operations per second for the queries assigned to a specific resource group. It allocates bandwidth, ensures the use of high-priority resource groups, and avoids excessive use of disk bandwidth. The value of the parameter is set on a per-tablespace basis. 

> **Note** Disk I/O limits are only available when you use Linux Control Groups v2. See [Configuring and Using Resource Groups](#topic71717999) for more information.

> When you limit disk I/O you speficy:

Reviewer: ```suggestion
When you limit disk I/O you specify:
```

- The tablespace name or the tablespace object ID (OID) you set the limits for. Use `*` to set limits for all tablespaces.

> - The values for `rpbs` and `wpbs` to limit the maximum read and write disk I/O throughput in the resource group, in MB/S. The default value is `max`, which means there is no limit. 

Reviewer: Typo.

```suggestion
- The values for `rbps` and `wbps` to limit the maximum read and write disk I/O throughput in the resource group, in MB/sec. The default value is `max`, which means there is no limit. 
```

One more question, we used the same syntax, but different unit. The cgroup v2 is using `byte/sec` as unit. But here, with the same name, we changed the unit to `MB/sec`? Is this intentional?

Author: Thanks for spotting the typo. I'll let @wenwangBJ @SmartKeyerror answer the question.

Reviewer 3: Yes, we change the Unit to MB/sec for user to configure it easier and more understandable.

And the in the view gp_toolkit.gp_resgroup_iostats_per_host  still keep the Unit B/sec which after discussion with dev engineers.
eg:postgres=# select * from gp_toolkit.gp_resgroup_iostats_per_host ; 
    rsgname    | hostname | tablespace |   rbps   |   wbps   | riops | wiops  
---------------+----------+------------+----------+----------+-------+------- 
admin_group   | stable   | pg_default | 21356347 | 29369067 |   162 |    36

- The values for `riops` and `wiops` to limit the maximum read and write I/O operations per second in the resource group. The default value is `max`, which means there is no limit. 

If the parameter `IO_LIMIT` is not set, the default value for `rbps`, `wpbs`, `riops`, and `wiops`s is set to `max`, which means that there are no disk I/O limits. In this scenario, the `gp_toolkit.gp_resgroup_config` system view displays its value as `-1`.

> 

Reviewer: Maybe we can add some examples here, especially for the semantically consistent syntax of this IO limit.

Or, do we actually have a syntax grammar here we can share?

```
# all table spaces with write throughput limit 1000MB/sec
*: wbps=1000

# pg_default table space with read throughput limit 1000MB/sec
pg_default: wbps=max, rbps=1000, wiops=max, riops=max
# simpler syntax (if supported)
pg_default: rbps=1000
```

Author: it should be on the CREATE/ALTER RESOURCE GROUP command reference pages, and down in the Examples section of this page too.

Reviewer: Yes, I saw those. Thanks.

## <a id="topic71717999"></a>Configuring and Using Resource Groups 

### <a id="topic833"></a>Prerequisites

> Greenplum Database resource groups use Linux Control Groups \(cgroups\) to manage CPU resources and disk I/O. There are two versions of cgroups: cgroup v1 and cgroup v2, which differ in the virtual file hierarchy implemented by v2. Greenplum Database 7 supports both versions, but it uses cgroup v2 by default, and it only supports the parameter `IO_LIMIT` for cgroup v2. The version of Linux Control Groups shipped by default with your Linux distribution depends on the operating system version. For Enterprise Linux 8 and older, the default version is v1. For Enterprise Linux 9 and later, the default version is v2. For detailed information about cgroups, refer to the Control Groups documentation for your Linux distribution.

Reviewer: Are we actually using `group_v2` as default? 
I believe we are using `none` as default.

A general comments, the v1 and v2 not only different from the file hierarchy, but also different on the controllers, constraints, delegation, security, etc. So, I just removed the sentence on hierarchy difference.

Author: @wenwangBJ @SmartKeyerror can you answer the question regarding default cgroup version used?

Author: @xinzweb I believe you made a suggested change but I cannot see it here.

Reviewer: ```suggestion
Greenplum Database resource groups use Linux Control Groups \(cgroups\) to manage CPU resources and disk I/O. There are two versions of cgroups: cgroup v1 and cgroup v2. Greenplum Database 7 supports both versions. It only supports the parameter `IO_LIMIT` for cgroup v2. The version of Linux Control Groups shipped by default with your Linux distribution depends on the operating system version. For Enterprise Linux 8 and older, the default version is v1. For Enterprise Linux 9 and later, the default version is v2. For detailed information about cgroups, refer to the Control Groups documentation for your Linux distribution.
```

Reviewer: Let me try again to see if you can see the above changes.

Author: I can see it now, let's see if the team can confirm about the default settings.

Reviewer 2: Since the Cgroup needs to be configured well if the customer wants to use resource group, we want them quick get started, so set the default value to `none`.

Reviewer 3: Now the default gp_resource_manager is none.

For cgroup v1, the output is `tmpfs`. For cgroup v2, output is `cgroup2fs`.

You do not need to change your version of cgroup, you can simply skip to [Configuring cgroup v1](#cgroupv1) or [Configuring cgroup v2](#cgroupv2) in order to complete the configuration prerequisites. However, if you prefer to switch from cgroup v1 to v2, run the following commands as root:

- Red Hat 8/Rocky 8/Oracle 8 systems:
    ```
    grubby --update-kernel=/boot/vmlinuz-$(uname -r) --args="systemd.unified_cgroup_hierarchy=1".
    ```
- Ubuntu systems:
    ```
    vim /etc/default/grub
    # add or modify: GRUB_CMDLINE_LINUX="systemd.unified_cgroup_hierarchy=1"
    update-grub && reboot now
    ```

If you want to switch from cgroup v2 to v1, run the following commands as root:

- Red Hat 8/Rocky 8/Oracle 8 systems:
    ```
    grubby --update-kernel=/boot/vmlinuz-$(uname -r) --args="systemd.unified_cgroup_hierarchy=0 systemd.legacy_systemd_cgroup_controller"
    ```
- Ubuntu systems:
    ```
    vim /etc/default/grub
    # add or modify: GRUB_CMDLINE_LINUX="systemd.unified_cgroup_hierarchy=0"
    update-grub && reboot now
    ```

> After that, reboot your host and make the changes take effect.

Reviewer: Well, we already add `reboot now` for the `Ubuntu` system, then in this case, do we still ask them to `reboot`? 

If we ask them to reboot, maybe we don't need to add the `&& reboot now` above?

Reviewer 3: @mperezfuster I think we have sync it for several rounds, please update all the relative contexts follow the PR:  https://github.com/greenplum-db/gpdb/pull/16327/files

    grubby --update-kernel=/boot/vmlinuz-$(uname -r) --args="systemd.unified_cgroup_hierarchy=0 systemd.legacy_systemd_cgroup_controller"
    ```
- Ubuntu systems:
    ```
    sudo vim /etc/default/grub
    add or modify: GRUB_CMDLINE_LINUX="systemd.unified_cgroup_hierarchy=0"
    sudo update-grub
    vim /etc/default/grub
    # add or modify: GRUB_CMDLINE_LINUX="systemd.unified_cgroup_hierarchy=0"
    update-grub
    ```

After that, reboot your host and make the changes take effect.
After that, reboot your host in order for the changes to take effect.

#### <a id="cgroupv1"></a>Configuring cgroup v1

Complete the following tasks on each node in your Greenplum Database cluster to set up cgroups v1 for use with resource groups:

1. If not already installed, install the Control Groups operating system package on each Greenplum Database node. The command that you run to perform this task will differ based on the operating system installed on the node. You must be the superuser or have `sudo` access to run the command:
1. Locate the cgroups configuration file `/etc/cgconfig.conf`. You must be the superuser or have `sudo` access to edit this file:

    ```

>     sudo yum install libcgroup-tools

Reviewer 2: we need the `libcgroup-tools` to use the config file to control the creation of Cgroup dirs.

Author: @SmartKeyerror this is already stated on the Platform Requirements page when installing Greenplum: https://docs.vmware.com/en/VMware-Greenplum/7/greenplum-database/install_guide-platform-requirements-overview.html#operating-systems

    ```
    cgconfigparser -l /etc/cgconfig.conf 
    ```

1. To automatically recreate Greenplum Database required cgroup hierarchies and parameters when your system is restarted, configure your system to enable the Linux cgroup service daemon `cgconfig.service` at node start-up. To ensure the configuration is persisten after reboot, run the following commands as user root. For Redhat/Oracle/Rocky 8.x and older:

    ```
    systemctl enable cgconfig.service
    ```

    To start the service immediately \(without having to reboot\) enter:

    ```
    systemctl start cgconfig.service
    ```

1. Identify the `cgroup` directory mount point for the node. For Redhat/Oracle/Rocky 8.x and older, run the following as root:

    ```
    grep cgroup /proc/mounts
    ```

    The first line of output identifies the `cgroup` mount point.

1. Verify that you set up the Greenplum Database cgroups configuration correctly by running the following commands. Replace \<cgroup\_mount\_point\> with the mount point that you identified in the previous step:

    ```
    ls -l <cgroup_mount_point>/cpu/gpdb

>     ls -l <cgroup_mount_point>/cpuacct/gpdb

Reviewer: we don't need this one. both `cpu` and `cpuacct` are pointing to the same location `cpu,cpuacct`.

Author: @SmartKeyerror can you confirm?

Reviewer 2: Yeah, that's right, we just need to verify `<cgroup_mount_point>/cpu/gpdb`

Reviewer 3: I think we have sync it for several rounds, please update all the relative contexts follow the PR: https://github.com/greenplum-db/gpdb/pull/16327/files

    ls -l <cgroup_mount_point>/cpuset/gpdb
    ls -l <cgroup_mount_point>/memory/gpdb
    ```

>     If these directories exist and are owned by `gpadmin:gpadmin`, you have successfully configured cgroups for Greenplum Database CPU resource management.

Reviewer: ```suggestion
    If these directories exist and are owned by `gpadmin:gpadmin`, you have successfully configured cgroups for Greenplum Database resource management.
```

I removed the `CPU`, since we also covered `memory` here (even though just for the reading purpose)

#### <a id="cgroupv2"></a>Configuring cgroup v2

1. Configure the system to mount `cgroups-v2` by default during system boot by the `systemd` system and service manager as user root.

    ```
    grubby --update-kernel=ALL --args="systemd.unified_cgroup_hierarchy=1"
    ```

1. Reboot the system for the changes to take effect.
    ```
    reboot now
    ```
1. Create the directory `/sys/fs/cgroup/gpdb` and ensure `gpadmin` user has read and write permission on it.
    ```
    mkdir -p /sys/fs/cgroup/gpdb
    chown -R gpadmin:gpadmin /sys/fs/cgroup/gpdb
    ```
1. Ensure that `gpadmin` has write permission on `/sys/fs/cgroup/cgroup.procs`.
    ```
    usermod -aG root gpadmin
    chmod g+w /sys/fs/cgroup/cgroup.procs

>     ```

Reviewer: Shall we add the reason why we need this permission at the top level?

I am thinking something like

```
We have to enable gpadmin to access top level `cgroup.procs` since we need to move the Greenplum processes from user slices to `/sys/fs/cgroup/gpdb/` after cluster started in order to manage the postmaster services and all its auxiliary processes like checkpoint, bgwriter, wal writer etc.
```

Author: @wenwangBJ @SmartKeyerror can you confirm?

Reviewer 2: agree w/ Shine's comments.

Reviewer 3: @mperezfuster  I think we have sync it for several rounds, please update all the relative contexts follow the PR: https://github.com/greenplum-db/gpdb/pull/16327/files

SELECT * FROM gp_toolkit.gp_resgroup_status;
```

### <a id="topic23a"></a>Viewing Resource Group CPU/Memory Usage Per Host 

> The [gp\_resgroup\_status\_per\_host](../ref_guide/system_catalogs/catalog_ref-views.html#gp_resgroup_status_per_host) `gp_toolkit` system view enables you to view the real-time CPU and memory usage of a resource group on a per-host basis. To view this information:

Reviewer 3: Please add the relative part to the document: gp_toolkit.gp_resgroup_status_per_segment

Reviewer 3: Update the example due to the PR revert that not have <bypass> lines anymore, if you needed:
postgres=# select * from gp_toolkit.gp_resgroup_status_per_segment ;

groupid | groupname | segment_id | vmem_usage

---------+---------------------+------------+------------

6438 | admin_group         |         -1 |          8 
6438 | admin_group         |          0 |        141 
6438 | admin_group         |          1 |        133 
6438 | admin_group         |          2 |        133
